scipy/signal/_spectral_py.py

```diff
@@ -85,71 +111,200 @@ def lombscargle(x,
 
-    Then plot the normalized periodogram:
+    Then plot the periodogram for each of the normalize options:
+
+    >>> ax_p.plot(w, pgram_power)
+    >>> ax_p.set_xlabel('Angular frequency [rad/s]')
+    >>> ax_p.set_ylabel('Power')
 
-    >>> ax_w.plot(w, pgram)
-    >>> ax_w.set_xlabel('Angular frequency [rad/s]')
-    >>> ax_w.set_ylabel('Normalized amplitude')
+    >>> ax_n.plot(w, pgram_norm)
+    >>> ax_n.set_xlabel('Angular frequency [rad/s]')
+    >>> ax_n.set_ylabel('Normalized')
+
+    >>> ax_a.plot(w, np.abs(pgram_amp))
+    >>> ax_a.set_xlabel('Angular frequency [rad/s]')
+    >>> ax_a.set_ylabel('Amplitude')
+
+    >>> plt.tight_layout()
     >>> plt.show()
 
     """
+
+    # if no weights are provided, assume all data points are equally important
+    if weights is None:
+        weights = np.ones_like(x, dtype=np.float64)
+
+    # validate input shapes
+    if x.ndim != 1:
+        raise ValueError("x array is not 1D")
+    if y.ndim != 1:
+        raise ValueError("y array is not 1D")
+    if freqs.ndim != 1:
+        raise ValueError("freqs array is not 1D")
+    if weights.ndim != 1:
+        raise ValueError("weights array is not 1D")
+    if x.shape != y.shape or x.shape != weights.shape:
+        raise ValueError("Input arrays do not have the same shape.")
+
+    # check for any freq == 0
+    if np.any(freqs == 0):
+        raise ZeroDivisionError("Freqs array contains a 0.")
+    
+    # check for weights summing to 0
+    if weights.sum() == 0:
+        raise ZeroDivisionError("Weights sum to 0.")
+    
+    # check for any negative weights
+    if np.any(weights < 0):
+        raise ValueError("Each weight must be >= 0.")
+
+    # validate normalize parameter
+    if isinstance(normalize, bool):
+        # if bool, convert to str literal
+        normalize = "normalize" if normalize else "power"
+    else:
+        # if neither a bool or str
+        if not isinstance(normalize, str):
+            raise TypeError("Normalize type must a bool or str.")
+
+    if normalize not in ["power", "normalize", "amplitude"]:
+        raise ValueError(
+            "Normalize must be: "
+            "False (or 'power'), "
+            "True (or 'normalize'), "
+            "or 'amplitude'."
+        )
+    
+    # for normalize == 'amplitude', check if freq_min < freq_critical
+    if (normalize == 'amplitude'):
+        freq_critical = 2*np.pi/(x.max() - x.min())  # rad/sec
+        if (freqs.min() < freq_critical):
+            message = (f"Frequencies below {freq_critical:.3f} rad/sec are longer than "
+                       "the provided time range, therefore their estimated amplitudes "
+                       "will tend towards infinity.")
+            warnings.warn(message, stacklevel=2)
+
+    # convert inputs to contiguous arrays with high precision
     x = np.ascontiguousarray(x, dtype=np.float64)
     y = np.ascontiguousarray(y, dtype=np.float64)
     freqs = np.ascontiguousarray(freqs, dtype=np.float64)
```

Comment:
It's a leftover. I do know that in some languages array types aren't guaranteed to be contiguous (I had to deal with this extensively in ObjC for an ever-growing array). But I have no idea when this is likely to be an issue with numpy. And I'm not familiar with the internals enough to know of a simple way to force the creation of non-contiguous array to test the performance impact.

Reply:
I am not too familiar with the internals either, but I am reasonably sure it is not an issue when having pure Python code.

Reply:
@DietBru I remember now also why I kept it, I wanted an explicit conversion to float64 to minimize the chance of numerical issues. I guess, for each array, I could test the type and convert if necessary. And it isn't clear that `ndarray.astype()` is the best solution. Do you have any thoughts on this?

Reply:
Done, but open to changes.

Reply:
I think it is reasonable to convert the input to `np.float64`, but it should be documented. I personally prefer raising an error over converting types, since it  opens up potential optimizations in the calling code.

There might be a niche use case, where a  `np.float32` or `np.float16` would be preferred: Evaluating if it makes sense to utilize a smaller float type for an optimized implementation (e.g., for a C implementation on an embedded device). 

Hence, an alternative would be to ensure that the input arrays are of the same float dtype. A more compact input validation could look like this (I think, it is ok to change the exception types):
```python
    if not (x.ndim == 1 and x.size > 0 and x.shape == y.shape == weights.shape):
        raise ValueError("Parameters x, y, weights must be 1d arrays of " +             
                         "equal non-zero length!")
    if not (freqs.ndim == 1 and freqs.size > 0):
        raise ValueError("Parameter freqs must be a 1d array of non-zero length!")
    if issubclass(x.dtype.type, np.integer) or np.iscomplexobj(x) or
       not (x.dtype == y.dtype == weights.dtype = freqs.dtype):
        raise ValueError("Parameters x, y, weights, freqs must be arrays of same " +
                         "float dtype (int and complex dtypes are not allowed)!")
    if not (np.all(weights >= 0) and np.sum(weights) > 0):
        raise ValueError("Parameter weights must have only non-negative entries " +
                         "which sum to a positive value!")
```

As said, raising an error and pulling the ``if`` statements together in the code is matter of personal taste - I do not have very strong opinions on that (but I think it is important to document internal type conversions).

Reply:
... just saw that the internal conversion is documented.